### PR TITLE
ARCH-1052 - Adding a workflow for unlocking terraform state

### PR DESCRIPTION
**Pull request by danielle-casella-adams**

### CODE REVIEW

File: workflow-templates/im-run-unlock-tf-state.yml [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+name: Unlock Terrafrom State
+on:
+  workflow_dispatch:
+    inputs:
+      lock-id:
+        description: The lock ID that needs to be removed.  Terraform provides this value (lock info.ID) in the message indicating there was an error acquiring the state lock.
+        required: true
+      environment:
+        description: The environment to deploy to - dev|qa|stage|prod|demo|uat
```
Comment (JosephDSchwartz) [inline workflow-templates/im-run-unlock-tf-state.yml:11]: The environment to deploy to? Should this be something like "The environment with the lock" or something like that?

File: workflow-templates/im-run-unlock-tf-state.yml [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    ]
+
+jobs:
+  set-vars:
+    runs-on: ubuntu-latest # Force this to run on github-hosted runner by using a tag that does not exist on self-hosted runners
+
+    outputs:
+      ENVIRONMENT: ${{ steps.clean-env.outputs.mapped_input }} # To use this value: ${{ needs.set-vars.outputs.ENVIRONMENT }}
+      RESOURCE_GROUP: ${{ steps.rgrp.outputs.RESOURCE_GROUP }} # To use this value: ${{ needs.set-vars.outputs.RESOURCE_GROUP }}
+      STORAGE_ACCOUNT: ${{ steps.stg-acct.outputs.STORAGE_ACCOUNT }} # To use this value: ${{ needs.set-vars.outputs.STORAGE_ACCOUNT }}
```
Comment (JosephDSchwartz) [inline workflow-templates/im-run-unlock-tf-state.yml:41]: Are these two leftovers from copy and paste? (I don't see them used below)

File: workflow-templates/im-run-unlock-tf-state.yml [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+              "uat": ["u", "uat"],
+              "prod": ["prod", "production", "p"]
+            }
+          error_on_no_match: true
+          custom_error_message: 'The environment must be Dev, QA, Stage Demo, UAT or Prod'
+
+      - run: echo "The current environment is ${{ steps.clean-env.outputs.mapped_input }}"
+
+  tf-unlock-state:
+    needs: [set-vars, validate-tag-for-prod-deploys]
```
Comment (JosephDSchwartz) [inline workflow-templates/im-run-unlock-tf-state.yml:68]: I don't see the `validate-tag-for-prod-deploys` job in this workflow.

File: workflow-templates/im-run-unlock-tf-state.yml [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    environment: ${{ needs.set-vars.outputs.ENVIRONMENT }}
+
+    defaults:
+      run:
+        shell: bash
+        working-directory: './${{ needs.set-vars.outputs.ENVIRONMENT }}' # TODO:  Verify this directory (./dev, ./qa, ./stage, ./prod, ./demo, ./uat) will be correct for your project
+
+    outputs:
+      tf_plan_name: ${{ steps.upload.outputs.tf_plan_name }}
+      maintenance_window_id: ${{ steps.open-window.outputs.maintenance-window-id }}
```
Comment (JosephDSchwartz) [inline workflow-templates/im-run-unlock-tf-state.yml:79]: Do we need these outputs?

File: workflow-templates/im-run-unlock-tf-state.yml [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+    steps:
+      - name: Checkout Repository
+        uses: actions/checkout@v2
+        with:
+          ref: ${{ env.GITHUB_REF }}
```
Comment (JosephDSchwartz) [inline workflow-templates/im-run-unlock-tf-state.yml:85]: Does this need to be defined above? Or can it be removed?

**Comment by danielle-casella-adams** [comment]
Man....I took a bunch of stuff out and the undid it because I realized I needed one section.  I guess I didn't go back and fix the rest of it.  Thank you for reviewing. :)

File: workflow-templates/im-run-unlock-tf-state.yml (hunk 5 above)
Comment (danielle-casella-adams) [reply workflow-templates/im-run-unlock-tf-state.yml:85]: I think it needs to be above.  Initially I thought I wouldn't need it but we need to do an init before running the unlock command so I think we do need it as an input to the workflow and we can set an env variable like the other workflows.

**Comment by danielle-casella-adams** [comment]
@JosephDSchwartz this should be ready to look at again.

**Review by JosephDSchwartz — APPROVED** [review]
Looks all good to me now! Sorry about all of the questions, but I'm glad I caught some stuff 😃

**Comment by danielle-casella-adams** [comment]
I don't mind the questions at all, they were all valid and helped prevent more re-work after it was merged!

**Pull request merged**